### PR TITLE
chore: add devenv

### LINKED ISSUE #477: devenv
### Describe the chore

Our setup for new developers, especially in this repository, is not quite intuitive and requires direct communication to the teams.

__The repository should have a dx / development first approach__

To ease the onboarding process of new members and developers in our organization, `helpwave/service` will be the first repository of helpwave adopting devenv (Nix) for configuring and running local developer environments.

https://devenv.sh/

A rough overview of the steps a new developer needs to do in the new setup:
- Installing devenv (and Nix) https://devenv.sh/getting-started/
- Running `devenv up` for starting all of our services via Dapr. Including Postgres and Redis.
- Running a warm-up like command on the first run or after a `git checkout` for preparing the environment. For example, `devenv shell warmup` to run migrations etc.

Notes:
- Our "old way" should stay intact for a while after devenv is in place to give developers some time for migrating
  - I plan around four weeks for that phase. After that, the old structure should be cleaned up.
- For UNIX-like systems, devenv is not a problem.
- Windows user needs to run devenv via WSL2

### PR DESCRIPTION
**Pull request by MaxSchaefer**
This pull requests tackles DX problems. Especially our initial onboarding.
__To all reviewers: Please try this out without further explanation of my side, to end-to-end test the onboarding and give me feedback.__ The broader idea is already outlined in the issue, before rolling this out organization wide, we will discuss this at the latest in the next weekly. Our old workflow via Makefile is still functional so that the migration can be done over time.

## Which issues does this pull request close?
closes #477 

Open To-dos for later implementation
- I've used [air](https://github.com/cosmtrek/air) for hot-reloading the go services, seams like a proper fit. Maybe we are finding a non-go branded solution for this. There are many out there.
- Doing Dapr component registration so that Dapr finds the redis instance, for example. This is not required yet for every local development.

## [OPTIONAL] Give testing instructions to reviewers

Going through our new _Getting started_ guide!

### CODE REVIEW

**Comment by FoseFx** [comment]
What is `devenv up` supposed to do? Is it a one-time thing, or do I run it every time? I looks like it *should* start a database, but that's not staying alive
```
Building shell ...
14:47:22 system     | postgres.1 started (pid=96951)
14:47:22 system     | redis.1 started (pid=96953)
14:47:22 system     | services.1 started (pid=96955)
14:47:22 redis.1    | 96958:C 26 Oct 2023 14:47:22.772 # WARNING Memory overcommit must be enabled! Without it, a background save or replication may fail under low memory condition. Being disabled, it can also cause failures without low memory condition, see https://github.com/jemalloc/jemalloc/issues/1328. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
14:47:22 redis.1    | 96958:C 26 Oct 2023 14:47:22.772 * oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
14:47:22 redis.1    | 96958:C 26 Oct 2023 14:47:22.772 * Redis version=7.2.2, bits=64, commit=00000000, modified=0, pid=96958, just started
14:47:22 redis.1    | 96958:C 26 Oct 2023 14:47:22.772 * Configuration loaded
14:47:22 redis.1    | 96958:M 26 Oct 2023 14:47:22.772 * Increased maximum number of open files to 10032 (it was originally set to 1024).
14:47:22 redis.1    | 96958:M 26 Oct 2023 14:47:22.772 * monotonic clock: POSIX clock_gettime
14:47:22 redis.1    | 96958:M 26 Oct 2023 14:47:22.773 * Running mode=standalone, port=6379.
14:47:22 redis.1    | 96958:M 26 Oct 2023 14:47:22.774 * Server initialized
14:47:22 redis.1    | 96958:M 26 Oct 2023 14:47:22.774 * Loading RDB produced by version 7.2.2
14:47:22 redis.1    | 96958:M 26 Oct 2023 14:47:22.774 * RDB age 251 seconds
14:47:22 redis.1    | 96958:M 26 Oct 2023 14:47:22.774 * RDB memory usage when created 0.84 Mb
14:47:22 redis.1    | 96958:M 26 Oct 2023 14:47:22.774 * Done loading RDB, keys loaded: 0, keys expired: 0.
14:47:22 redis.1    | 96958:M 26 Oct 2023 14:47:22.774 * DB loaded from disk: 0.000 seconds
14:47:22 redis.1    | 96958:M 26 Oct 2023 14:47:22.774 * Ready to accept connections tcp
14:47:22 services.1 | Note: using sslmode=disable, only use this script with local databases!
14:47:22 postgres.1 | The files belonging to this database system will be owned by user "max".
14:47:22 postgres.1 | This user must also own the server process.
14:47:22 postgres.1 | 
14:47:22 postgres.1 | The database cluster will be initialized with locale "C".
14:47:22 postgres.1 | The default text search configuration will be set to "english".
14:47:22 postgres.1 | 
14:47:22 postgres.1 | Data page checksums are disabled.
14:47:22 postgres.1 | 
14:47:22 postgres.1 | creating directory /home/max/Documents/helpwave/services/.devenv/state/postgres ... ok
14:47:22 postgres.1 | creating subdirectories ... ok
14:47:22 postgres.1 | selecting dynamic shared memory implementation ... posix
14:47:22 postgres.1 | selecting default max_connections ... 100
14:47:22 services.1 | error: no migration found for version 18: read down for version 18 .: file does not existerror: no migration found for version 18: read down for version 18 .: file does not exist
14:47:22 system     | services.1 stopped (rc=1)
14:47:22 system     | sending SIGTERM to postgres.1 (pid 96951)
14:47:22 system     | sending SIGTERM to redis.1 (pid 96953)
14:47:22 redis.1    | 96958:signal-handler (1698324442) Received SIGTERM scheduling shutdown...
14:47:22 postgres.1 | selecting default shared_buffers ... 32MB
14:47:22 postgres.1 | selecting default time zone ... Europe/Berlin
14:47:22 postgres.1 | creating configuration files ... caught signal
14:47:22 postgres.1 | initdb: removing data directory "/home/max/Documents/helpwave/services/.devenv/state/postgres"
14:47:22 system     | postgres.1 stopped (rc=-15)
14:47:22 redis.1    | 96958:M 26 Oct 2023 14:47:22.875 * User requested shutdown...
14:47:22 redis.1    | 96958:M 26 Oct 2023 14:47:22.875 * Saving the final RDB snapshot before exiting.
14:47:22 redis.1    | 96958:M 26 Oct 2023 14:47:22.881 * DB saved on disk
14:47:22 redis.1    | 96958:M 26 Oct 2023 14:47:22.881 # Redis is now ready to exit, bye bye...
14:47:22 system     | redis.1 stopped (rc=-15)
```

Also, direnv spams my console when I enter the services directory
```
direnv: loading ~/Documents/helpwave/services/.envrc
direnv: loading https://raw.githubusercontent.com/cachix/devenv/d1f7b48e35e6dee421cfd0f51481d17f77586997/direnvrc (sha256-YBzqskFZxmNb3kYVoKD9ZixoPXJh1C9ZvTLGFRkauZ0=)
direnv: using devenv
direnv: using cached devenv shell
direnv: PS1 cannot be exported. For more information see https://github.com/direnv/direnv/wiki/PS1
direnv: export +C_INCLUDE_PATH +DEVENV_DOTFILE +DEVENV_PROFILE +DEVENV_ROOT +DEVENV_STATE +IN_NIX_SHELL +LD_LIBRARY_PATH +LIBRARY_PATH +LOCALE_ARCHIVE +PGDATA +PGHOST +PGPORT +PKG_CONFIG_PATH +PYTHONPATH +REDISDATA +XDG_CONFIG_DIRS +name ~GOPATH ~GOROOT ~PATH ~XDG_DATA_DIRS
```

Also, I think we should mention that it takes some time to build everything.
We should also mention that direnv will load the devenv shell every time you enter the directory.

It also asked me to `direnv allow` the env before I installed devenv (I had direnv installed before).

**Comment by MaxSchaefer** [comment]
@FoseFx thanks for testing. Seams like the `14:47:22 postgres.1 | creating configuration files ... caught signal` failed. Could you try removing the `/.devenv` directory and run `devenv up` again? `devenv up` should bring up the whole environment like infra and hot-reloading for our services. You need to run this command every time you want to develop on this repository. direnv is a little bit noisy and not covered right now by me, I don't have it installed. I guess we could just get rid of the `.envrc` file.

@(no ping)Friedjof is testing right now (on NixOS). Let's jump on a call after he tries the setup as well to get rid of some problems.

Here is a little bit of my log of a successful run of `devenv up`
```
21:57:30 postgres.1 | The files belonging to this database system will be owned by user "maxrs".
21:57:30 postgres.1 | This user must also own the server process.
21:57:30 postgres.1 |
21:57:30 postgres.1 | The database cluster will be initialized with locale "C".
21:57:30 postgres.1 | The default text search configuration will be set to "english".
21:57:30 postgres.1 |
21:57:30 postgres.1 | Data page checksums are disabled.
21:57:30 postgres.1 |
21:57:30 postgres.1 | creating directory /Users/maxrs/Projects/helpwave/helpwave-services/.devenv/state/postgres ... ok
21:57:30 postgres.1 | creating subdirectories ... ok
21:57:30 postgres.1 | selecting dynamic shared memory implementation ... posix
21:57:31 postgres.1 | selecting default max_connections ... 100
21:57:31 postgres.1 | selecting default shared_buffers ... 128MB
21:57:31 postgres.1 | selecting default time zone ... Europe/Berlin
21:57:31 postgres.1 | creating configuration files ... ok
21:57:31 postgres.1 | running bootstrap script ... ok
21:57:32 postgres.1 | performing post-bootstrap initialization ... ok
21:57:32 postgres.1 | syncing data to disk ... ok
21:57:32 postgres.1 |
21:57:32 postgres.1 | initdb: warning: enabling "trust" authentication for local connections
21:57:32 postgres.1 | You can change this by editing pg_hba.conf or using the option -A, or
21:57:32 postgres.1 | --auth-local and --auth-host, the next time you run initdb.
21:57:32 postgres.1 |
21:57:32 postgres.1 | Success. You can now start the database server using:
21:57:32 postgres.1 |
21:57:32 postgres.1 |     pg_ctl -D /Users/maxrs/Projects/helpwave/helpwave-services/.devenv/state/postgres -l logfile start
21:57:32 postgres.1 |
21:57:32 postgres.1 |
21:57:32 postgres.1 | PostgreSQL initdb process complete.
21:57:32 postgres.1 |
21:57:32 postgres.1 |
21:57:32 postgres.1 | PostgreSQL is setting up the initial database.
21:57:32 postgres.1 |
21:57:32 postgres.1 | waiting for server to start....2023-10-26 19:57:32.440 GMT [59993] LOG:  starting PostgreSQL 14.9 on x86_64-apple-darwin22.5.0, compiled by clang version 11.1.0, 64-bit
21:57:32 postgres.1 | 2023-10-26 19:57:32.441 GMT [59993] LOG:  listening on Unix socket "/Users/maxrs/Projects/helpwave/helpwave-services/.devenv/state/pg-init-pJatV0/.s.PGSQL.5432"
21:57:32 postgres.1 | 2023-10-26 19:57:32.449 GMT [59994] LOG:  database system was shut down at 2023-10-26 19:57:32 GMT
21:57:32 postgres.1 | 2023-10-26 19:57:32.453 GMT [59993] LOG:  database system is ready to accept connections
21:57:32 postgres.1 |  done
21:57:32 postgres.1 | server started
```

**Comment by use-to** [comment]
@tanneberger

**Comment by PaulKalho** [comment]
After installing nix and devenv, `devenv up` will not successfully run.

```
Building shell ...
15:37:50 system     | redis.1 started (pid=101609)
15:37:50 system     | postgres.1 started (pid=101610)
15:37:50 system     | services.1 started (pid=101613)
15:37:50 redis.1    | 101614:C 28 Oct 2023 15:37:50.407 # WARNING Memory overcommit must be enabled! Without it, a background save or replication may fail under low memory condition. Being disabled, it can also cause failures without low memory condition, see https://github.com/jemalloc/jemalloc/issues/1328. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
15:37:50 redis.1    | 101614:C 28 Oct 2023 15:37:50.407 * oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
15:37:50 redis.1    | 101614:C 28 Oct 2023 15:37:50.407 * Redis version=7.2.2, bits=64, commit=00000000, modified=0, pid=101614, just started
15:37:50 redis.1    | 101614:C 28 Oct 2023 15:37:50.407 * Configuration loaded
15:37:50 redis.1    | 101614:M 28 Oct 2023 15:37:50.407 # Failed to configure LOCALE for invalid locale name.
15:37:50 system     | redis.1 stopped (rc=1)
15:37:50 system     | sending SIGTERM to postgres.1 (pid 101610)
15:37:50 system     | sending SIGTERM to services.1 (pid 101613)
15:37:50 system     | postgres.1 stopped (rc=-15)
15:37:50 services.1 | Terminated
15:37:50 services.1 | /nix/store/jrpjrgyphy59zkj1rhzwfdk6r06m6bm2-services: line 5: pid: unbound variable
15:37:50 system     | services.1 stopped (rc=1)
```

I also tried deleting  `.devenv`, same problem after that...

File: .envrc [hunk 1]
```diff
@@ -0,0 +1,3 @@
+source_url "https://raw.githubusercontent.com/cachix/devenv/d1f7b48e35e6dee421cfd0f51481d17f77586997/direnvrc" "sha256-YBzqskFZxmNb3kYVoKD9ZixoPXJh1C9ZvTLGFRkauZ0="
```
Comment (tanneberger) [inline .envrc:1]: is it necessary to pin(with revision and sha hash) the script here ?

**Review by tanneberger — APPROVED** [review]
Looks Good and Works on my site

Just one tiny note how you also could do this alternatively: I also use nixos for my deployment machines, so I write myself options for my NixOS services (see [here](https://github.com/tlm-solutions/datacare/blob/master/nixos-module/default.nix)) and then I can use these options in building qemu vms where I can then play around in (e.g. [vm-config](https://github.com/tlm-solutions/datacare/blob/master/tests/vm/default.nix))

File: .envrc (hunk 1 above)
Comment (MaxSchaefer) [reply .envrc:1]: Thanks for the review :) Yup, this is used by direnv.

File: nix/devenv/scripts.nix [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+      fi
+    '';
+  };
+
+  processes = {
+    services.exec = ''
+      set -euo pipefail
+
+      cleanup() {
+        kill -TERM $pid
```
Comment (FoseFx) [inline nix/devenv/scripts.nix:81]: looks like `$pid` is not set at https://github.com/helpwave/services/pull/478#issuecomment-1783816481

Comment (PaulKalho) [reply nix/devenv/scripts.nix:81]: Thank you,  actually the fix was to change my system locale. Somehow redis had a problem with a german locale. After changing to en_US and deleting .devenv it works like a charm!

**Review by PaulKalho — APPROVED** [review]

**Comment by use-to** [comment]
Looks like a working version. Probably some problems/improvements will appear after some time. But let's get this version on the road. 👍🏻

**Pull request merged**